be2-scala/src/main/scala/ch/epfl/pop/decentralized/GossipManager.scala

```diff
@@ -0,0 +1,123 @@
+    // selects a random peer from remaining peers
+    val randomPeer = connectionMediator ? ConnectionMediator.GetRandomPeer(activeGossip.map(_._1))
+    Await.result(randomPeer, duration) match {
+      // updates the list based on response
+      // if some peers are available we send
+      case ConnectionMediator.GetRandomPeerAck(serverRef, greetServer) =>
+        val alreadySent: List[ServerInfos] = activeGossip :+ (serverRef -> greetServer)
+        activeGossipProtocol += (rumorRpc -> alreadySent)
+        serverRef ! ClientAnswer(
+          Right(rumorRpc)
+        )
+      // else remove entry
+      case ConnectionMediator.NoPeer =>
+        activeGossipProtocol = activeGossipProtocol.removed(rumorRpc)
+    }
+  }
+
+  private def processResponse(response: JsonRpcResponse): Unit = {
+    val activeGossipPeers = activeGossipProtocol.filter((k, _) => k.id == response.id)
+    // response is expected because only one entry exists
+    if (activeGossipPeers.size == 1) {
+      activeGossipPeers.foreach { (rumorRpc, _) =>
+        if (response.result.isEmpty && Random.nextDouble() < stopProbability) {
+          activeGossipProtocol -= rumorRpc
+        } else {
+          sendRumorToRandomPeer(rumorRpc)
+        }
+      }
+    }
+  }
+
+  override def receive: Receive = {
+    case GossipManager.SendRumorToRandomPeer(rumorRpc) =>
+      sendRumorToRandomPeer(rumorRpc)
+
+    case GossipManager.ProcessResponse(jsonRpcResponse) =>
+      processResponse(jsonRpcResponse)
+  }
+
+}
+
+object GossipManager extends AskPatternConstants {
+  def props(dbActorRef: AskableActorRef, monitorRef: ActorRef, connectionMediatorRef: AskableActorRef): Props =
+    Props(new GossipManager(dbActorRef, monitorRef, connectionMediatorRef))
+
+  def gossipHandler(gossipManager: AskableActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map {
+    case Right(jsonRpcRequest: JsonRpcRequest) =>
+      jsonRpcRequest.method match
+        case MethodType.rumor =>
+          gossipManager ? SendRumorToRandomPeer(jsonRpcRequest)
+          Right(jsonRpcRequest)
+        case _ => Left(PipelineError(ErrorCodes.SERVER_ERROR.id, "GossipManager received a non expected jsonRpcRequest", jsonRpcRequest.id))
+    case graphMessage @ _ => Left(PipelineError(ErrorCodes.SERVER_ERROR.id, "GossipManager received an unexpected message:" + graphMessage, None))
+  }
+
+  def monitorResponse(gossipManager: AskableActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map {
+    case Right(jsonRpcResponse: JsonRpcResponse) =>
+      gossipManager ? ProcessResponse(jsonRpcResponse)
+      Right(jsonRpcResponse)
+    case graphMessage @ _ => graphMessage
+  }
+
+  sealed trait Event
+  final case class MonitoredRumor(jsonRpcRumor: JsonRpcRequest)
+  final case class SendRumorToRandomPeer(jsonRpcRequest: JsonRpcRequest)
+  final case class ProcessResponse(jsonRpcResponse: JsonRpcResponse)
```

Comment:
I find that name a bit confusing, its going to clash with the code that process the messages in the rumors I think